### ISSUE TITLE
createRouterTransport with Jest's jsdom raises error

### ISSUE THREAD

**Issue by timostamm** [connectrpc/connect-es#658]
**Describe the bug**

When using `createRouterTransport()` to mock an API, tests running with Jest in the `jsdom` environment can raise the following errors:

1. ReferenceError: TextEncoder is not defined
2. iterable[Symbol.asyncIterator] is not a function

This first error is because jsdom does not support TextEncoder and TextDecoder, see https://github.com/jsdom/jsdom/issues/2524. The second error is because we use `instanceof Uint8Array` in the code base to narrow down types, and Jest breaks it, see https://github.com/jestjs/jest/issues/4422. 

**To Reproduce**

Get [this gist](https://gist.github.com/timostamm/f4a9eb117c237f6a2dfec0567869c5f1), run `npm ci` and `npm test`. The tests pass, but only because `jest.config.js` contains the following setting:

```ts
  globals: {
    TextEncoder: TextEncoder,
    TextDecoder: TextDecoder,
    Uint8Array: Uint8Array,
  }
```

Removing the globals produces the errors described above.

**Environment (please complete the following information):**
- @bufbuild/connect 0.9.0
- Jest 29.5.0
- Node.js: 18.15.0

**Additional context**

jsdom is  - quoting the repository description - "a JavaScript implementation of various web standards, for use with Node.js". 

It seems odd that using it in Jest introduces incompatibility that wouldn't otherwise exist: `Uint8Array` is available since Node.js from 2013 and IE 10 from 2012. `TextEncoder` is available since Node.js from 2017 (but can easily be implemented with Node's `Buffer`), and the only browser that does not implement it is IE 11 from 2013.

**Comment by nicole0707** [comment]
A new issue on `Uint8Array`. If test files are running in sequence, they are working well with globals setting in `jest.config.js`.  However it complains `iterable[Symbol.asyncIterator] is not a function
` again when in parallel.

I suspect it's somehow changed during running and result in remaining files can't access it correctly.  We can potentially see that from [`globals`](https://jestjs.io/docs/configuration#globals-object) doc in jest
```
If you specify a global reference value (like an object or array) here, and some code mutates that value in the midst of running a test, that mutation will not be persisted across test runs for other test files
``` 

So I setup an custom environment to reset `Uint8Array` before each test file running to fix that.

```ts
class CustomTestEnvironment extends TestEnvironment {
  async setup() {
    await super.setup();
    /**
     * Set TextEncoder/TextDecoder if undefined
     */
    if (this.global.TextEncoder === undefined) {
      this.global.TextEncoder = TextEncoder;
    }

    if (this.global.TextDecoder === undefined) {
      (this.global as any).TextDecoder = TextDecoder;
    }
    /**
     * Uint8Array configured in globals could be changed when running tests in parallel
     * Reset Uint8Array before each test file is running
     */
    this.global.Uint8Array = Uint8Array;
  }
}
```

**Comment by timostamm** [comment]
Good catch, @nicole0707! 

`Uint8Array` is a constructor function that does not hold state. But if the issue comes back with parallel runs, obviously _something_ is going wrong, it might just be more complex than expected. 

The docs also say:

> In addition, the globals object must be json-serializable, so it can't be used to specify global functions.

Jest is most likely making a copy of `globals` to prevent individual tests from modifying it. I imagine this can break in mysterious ways with the constructor functions. Your approach of making them available through the environment seems to be the way to go.

**Comment by timostamm** [comment]
> Jest is most likely making a copy of `globals` to prevent individual tests from modifying it. 

@dimitropoulos' test seems to confirm that in https://github.com/bufbuild/connect-query/pull/50#issuecomment-1568411746.

We will look into publishing a test environment as an npm package, so that it is easy to set up tests.
